### PR TITLE
macOS Unified Access Logging (UAL)

### PR DESCRIPTION
**Pull request by dabresua**
This PR completes the work done by @puffyCid in [https://github.com/osquery/osquery/pull/7259](https://github.com/osquery/osquery/pull/7259) which also continued the work made by @GarretReece and @mike-myers-tob.

The funcionality enables osquery to read from UAL using its API.

My contribution is explained in the blueprint [https://github.com/osquery/osquery/issues/7591](https://github.com/osquery/osquery/issues/7591) which, in summary:

* Limits the rows the table returns using a flag (ual_max_rows), 100 by default
* If the user launches a non-specific query ```select * from unified_log``` returns the rows continuing from the last one, so user can extract all the logs by querying multiple times.

Adds unitary tests for the flag and for the sequential extraction.

Closes:
* [https://github.com/osquery/osquery/issues/7591](https://github.com/osquery/osquery/issues/7591)
* [https://github.com/osquery/osquery/pull/7259](https://github.com/osquery/osquery/pull/7259)
* [https://github.com/osquery/osquery/pull/6904](https://github.com/osquery/osquery/pull/6904)
* [https://github.com/osquery/osquery/issues/5760](https://github.com/osquery/osquery/issues/5760)

Co-authored-by: puffycyan@gmail.com
Co-authored-by: GarretReece@users.noreply.github.com

### CODE REVIEW

**Comment by dabresua** [comment]
The last changes implement what we discussed at the [osquery office hours 2022-05-24](https://hackmd.io/7aYrhwjQSrmlMhp_3fyS9A).

Getting data sequentially is now triggered by the command `select * from unified_log where timestamp > -1`. I've tried to implement a sentinel value like `LAST_TIMESTAMP`, but I found the SQLite engine was not expecting a non-numerical value and rejected it, returning nothing.

Instead of a flag, a HIDDEN column is used for max_rows. The default value is 100, the way to modify it is by adding the clause `max_rows = N`. For example, `select * from unified_logs where max_rows = 123` returns up to 123 rows.

User shall combine both clauses, i.e. `select * from unified logs where timestamp > -1 and max_rows = 1000`. This query returns the first 1000 log entries, and if we run it again, it will return the next 1000 log entries ordered by timestamp.

**Comment by dabresua** [comment]
Finishes the work done in #7259

**Comment by mike-myers-tob** [comment]
@directionless is going to re-review but it's moving to the 5.5 milestone

File: specs/darwin/unified_log.table [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  Column("message", TEXT, "composed message", additional=True),
+  Column("activity", BIGINT, "the activity ID associate with the entry", additional=True),
+  Column("process", TEXT, "the name of the process that made the entry", additional=True),
+  Column("pid", BIGINT, "the pid of the process that made the entry", additional=True),
+  Column("sender", TEXT, "the name of the binary image that made the entry", additional=True),
+  Column("tid", BIGINT, "the tid of the thread that made the entry", additional=True),
+  Column("category", TEXT, "the category of the os_log_t used", additional=True),
+  Column("subsystem", TEXT, "the subsystem of the os_log_t used", additional=True),
+  Column("level", TEXT, "the severity level of the entry"),
+  Column("max_rows", INTEGER, "the max number of rows returned", additional=True, hidden=True)
```
Comment (directionless) [inline specs/darwin/unified_log.table:20]:
```suggestion
  Column("max_rows", INTEGER, "the max number of rows returned (defaults to 100)", additional=True, hidden=True)
```

File: specs/darwin/unified_log.table [hunk 2]
```diff
@@ -0,0 +1,28 @@
+table_name("unified_log")
+description("Queries the OSLog framework for entries in the system log. "
+            "The maximum number of rows returned is limited for performance issues. "
+            "The condition \"max_rows = N\" sets the maximum number of rows returned (100 by default). "
```
Comment (directionless) [inline specs/darwin/unified_log.table:4]:
```suggestion
```

File: specs/darwin/unified_log.table [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            "This table introduces a new idiom for extracting sequential data in batches using multiple queries, ordered by timestamp. "
+            "To trigger it, the user should include the condition \"timestamp > -1\".")
```
Comment (directionless) [inline specs/darwin/unified_log.table:6]:
```suggestion
            "To trigger it, the user should include the condition \"timestamp > -1\", and the table will handle pagination.")
```

File: osquery/tables/system/darwin/unified_log.mm [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+                                                      {"sender", false},
+                                                      {"tid", true},
+                                                      {"subsystem", false},
+                                                      {"category", false}};
+
+/**
+ * @brief The backing store keys for saving the lst data extracted.
+ */
+const std::string kUALtimestampKey{"ual_timestamp"};
+const std::string kUALcountKey{"ual_counter"};
```
Comment (directionless) [inline osquery/tables/system/darwin/unified_log.mm:60]: Is it ual, or aul?

File: osquery/tables/system/darwin/unified_log.mm [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+};
+
+void SequentialContext::load() {
+  std::string str;
+  auto s = getDatabaseValue(kPersistentSettings, kUALtimestampKey, str);
+  if (s.ok())
+    timestamp = std::stod(str);
+  s = getDatabaseValue(kPersistentSettings, kUALcountKey, str);
+  if (s.ok())
+    count = std::stod(str);
```
Comment (directionless) [inline osquery/tables/system/darwin/unified_log.mm:107]: `count` is an int. Shouldn't be `std::stod`, right?

**Review by directionless — COMMENTED** [review]
Code looks mostly fine to me. @alessandrogario @Smjert either you look?

One question -- I think this has a single tracking state for the last seen state. So if you had two queries each using this table and the `-1` it wouldn't work. Correct?

**Comment by directionless** [comment]
I tried `select timestamp, process, substr(message, 0, 200) from unified_log where max_rows = 50 and timestamp = -1;` and it returned no data and mostly wedged.

File: osquery/tables/system/darwin/unified_log.mm (hunk 4 above)
Comment (directionless) [reply osquery/tables/system/darwin/unified_log.mm:60]: I think you have it correct.

**Comment by directionless** [comment]
Ha. Correcting myself, `where timestamp > -1` works correctly. 

And `where timestamp = -1` seems like a footgun we probably shouldn't try to fix?

File: osquery/tables/system/darwin/unified_log.mm (hunk 5 above)
Comment (dabresua) [reply osquery/tables/system/darwin/unified_log.mm:107]: Yes, should use `std::stoi` instead

**Comment by dabresua** [comment]
> Ha. Correcting myself, `where timestamp > -1` works correctly.
> 
> And `where timestamp = -1` seems like a footgun we probably shouldn't try to fix?

As you said before, SQLite allows only the data with `timestamp = -1`, so it will be filtered. However, someway the UAL engine gets stuck, I'm adding an escape condition for that situation.

**Comment by directionless** [comment]
I think this is probably mergable? At least, I don't think I have other comments. @Smjert, @alessandrogario Either of you have comments?

@dabresua I don't know if it makes sense, you could consider tossing in the github `Co-authored-by` for the other folks. Or not.

**Review by directionless — APPROVED** [review]
Sounds like we all think this is approvable. wooo! 

Thank you so much for sticking with it @dabresua it's a huge PR and improvement

**Pull request merged**